streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java

```diff
@@ -167,7 +169,7 @@ private void updateHead() {
 
             final long timestamp;
             try {
-                timestamp = timestampExtractor.extract(deserialized, timestamp());
+                timestamp = timestampExtractor.extract(deserialized, partitionTime);
```

Comment:
Can we also update method `timestamp()` to `headRecordTimestamp()` to be more explicit what it returns? It's orthogonal to the actual fix, but might be a good improvement.

Reply:
Similar, can we piggy-back some cleanup to `PartitionGroup` ?

```
- rename PartitionGroup#timestamp() to PartitionGroup#streamTime()
(also update the JavaDocs, that seems to be wrong)

- in `clear()` reset `streamTime` to UNKNOWN ?

- in `nextRecord()`: do we need to check if `queue != null` and do we need to check if `record != null` (seem it's ensure that both can never be `null` ?)
```

Reply:
Ack to all...except the last point. We do check both for null..?

Reply:
Seems, we check both for `null` atm:
```
        final RecordQueue queue = nonEmptyQueuesByTime.poll();
        info.queue = queue;

        if (queue != null) {
            // get the first record from this queue.
            record = queue.poll();

            if (record != null) {
                --totalBuffered;
```

But I think they cannot be `null`, could they?

Reply:
Sorry, misunderstood your question. Yes, either one could potentially be null if we don't yet have new records to process?

Reply:
Good point. ` final RecordQueue queue = nonEmptyQueuesByTime.poll();` could return `null`. However, I am wondering if `record = queue.poll();` could return `null`, because it's called `nonEmptyQueuesByTime` -- hence, `queue` should never be empty?

Reply:
I think I agree the second null check should never happen.